File: circleindexpool.go
package circleindexpool

type CircleIndexItem struct {
	nIndex int
	pNext  *CircleIndexItem
}

type CircleIndexPool struct {
	nIndexUsedCount  int
	nIndexTotalCount int
	pIndexPool       []CircleIndexItem
	ppIndexUsedPool  []*CircleIndexItem
	pHeadItem        *CircleIndexItem
	pTailItem        *CircleIndexItem
}

func NewCircleIndexPool(count int) *CircleIndexPool {
	ins := &CircleIndexPool{}
	ins.Init(count)
	return ins
}

func (this *CircleIndexPool) Init(count int) bool {
	if count < 0 ||
		this.nIndexTotalCount != 0 {
		return false
	}

	this.nIndexTotalCount = count
	this.pIndexPool = make([]CircleIndexItem, count)
	//	init the pool
	for i, _ := range this.pIndexPool {
		this.pIndexPool[i].nIndex = i + 1
		if i == count-1 {
			//	last one
			this.pIndexPool[i].pNext = nil
		} else {
			this.pIndexPool[i].pNext = &this.pIndexPool[i+1]
		}
	}
	this.pHeadItem = &this.pIndexPool[0]
	this.pTailItem = &this.pIndexPool[count-1]
	this.ppIndexUsedPool = make([]*CircleIndexItem, count+1)

	return true
}

func (this *CircleIndexPool) Free() {
	this.nIndexTotalCount = 0
	this.nIndexUsedCount = 0

	if nil != this.pIndexPool {
		this.pIndexPool = nil
	}
	if nil != this.ppIndexUsedPool {
		this.ppIndexUsedPool = nil
	}

	this.pHeadItem = nil
	this.pTailItem = nil
}

func (this *CircleIndexPool) AllocateIndex() int {
	if nil == this.pHeadItem {
		return 0
	}

	pUsedItem := this.pHeadItem
	nIndex := this.pHeadItem.nIndex
	this.pHeadItem = this.pHeadItem.pNext

	if nil == this.pHeadItem {
		this.pTailItem = nil
	}

	this.ppIndexUsedPool[nIndex] = pUsedItem
	this.nIndexUsedCount++

	return nIndex
}

func (this *CircleIndexPool) FreeIndex(nIndex int) {
	pUsedItem := this.ppIndexUsedPool[nIndex]
	if nil == pUsedItem {
		return
	}

	this.ppIndexUsedPool[nIndex] = nil
	pUsedItem.pNext = nil

	if nil == this.pTailItem {
		this.pHeadItem = pUsedItem
		this.pTailItem = pUsedItem
	} else {
		this.pTailItem.pNext = pUsedItem
		this.pTailItem = pUsedItem
	}

	this.nIndexUsedCount--
}

func (this *CircleIndexPool) GetTotalIndexCount() int {
	return this.nIndexTotalCount
}

func (this *CircleIndexPool) GetUsedIndexCount() int {
	return this.nIndexUsedCount
}

File: CircleIndexPool.cpp
#include "CircleIndexPool.h"
#include <Windows.h>
//////////////////////////////////////////////////////////////////////////
CircelIndexPool::CircelIndexPool()
{
	m_nIndexTotalCount = 0;
	m_nIndexUsedCount = 0;
	m_pIndexPool = NULL;
	m_ppIndexUsedPool = NULL;
	m_pHeadItem = m_pTailItem = NULL;
}

CircelIndexPool::~CircelIndexPool()
{
	Free();
}

bool CircelIndexPool::Init(int _nCount)
{
	if(m_nIndexTotalCount != 0 ||
		_nCount < 0)
	{
		return false;
	}

	m_nIndexTotalCount = _nCount;

	m_pIndexPool = new CircelIndexItem[_nCount];
	for(int i = 0; i < _nCount; ++i)
	{
		m_pIndexPool[i].nIndex = i + 1;
		if(i == _nCount - 1)
		{
			//	last one
			m_pIndexPool[i].pNext = NULL;
		}
		else
		{
			m_pIndexPool[i].pNext = m_pIndexPool + i + 1;
		}
	}

	//	get head and tail item
	m_pHeadItem = &m_pIndexPool[0];
	m_pTailItem = &m_pIndexPool[m_nIndexTotalCount - 1];

	m_ppIndexUsedPool = new CircelIndexItem*[_nCount + 1];
	memset(m_ppIndexUsedPool, 0, sizeof(CircelIndexItem*) * (_nCount + 1));

	return true;
}

void CircelIndexPool::Free()
{
	m_nIndexTotalCount = m_nIndexUsedCount = 0;

	if(NULL != m_pIndexPool)
	{
		delete [] m_pIndexPool;
		m_pIndexPool = NULL;
	}
	if(NULL != m_ppIndexUsedPool)
	{
		delete [] m_ppIndexUsedPool;
		m_ppIndexUsedPool = NULL;
	}

	m_pHeadItem = m_pTailItem = NULL;
}

int CircelIndexPool::AllocateIndex()
{
	if(NULL == m_pHeadItem)
	{
		return 0;
	}

	CircelIndexItem* pUsedItem = m_pHeadItem;
	int nIndex = m_pHeadItem->nIndex;
	m_pHeadItem = m_pHeadItem->pNext;

	if(NULL == m_pHeadItem)
	{
		//	already reach tail
		m_pTailItem = NULL;
	}

	//	push to used index pool
	m_ppIndexUsedPool[nIndex] = pUsedItem;
	++m_nIndexUsedCount;

	return nIndex;
}

void CircelIndexPool::FreeIndex(int _nIndex)
{
	//	get the used index
	CircelIndexItem* pUsedItem = m_ppIndexUsedPool[_nIndex];
	if(NULL == pUsedItem)
	{
		return;
	}

	m_ppIndexUsedPool[_nIndex] = NULL;
	pUsedItem->pNext = NULL;

	if(NULL == m_pTailItem)
	{
		m_pHeadItem = pUsedItem;
		m_pTailItem = pUsedItem;
	}
	else
	{
		//	add to tail
		m_pTailItem->pNext = pUsedItem;
		m_pTailItem = pUsedItem;
	}

	--m_nIndexUsedCount;
}